National Library of Norway.js

```diff
@@ -0,0 +1,205 @@
+	GNU Affero General Public License for more details.
+
+	You should have received a copy of the GNU Affero General Public License
+	along with Zotero. If not, see <http://www.gnu.org/licenses/>.
+
+	***** END LICENSE BLOCK *****
+*/
+
+function detectWeb(doc, url) {
+	if (url.match("/nbsok/search?")) {
+		return "multiple";
+	} else if (url.match("/nbsok/nb/")) {
+		var mediaTypes = {
+			'Bøker': 'book',
+			'Tidsskrift': 'book', // complete issues or volumes, not articles
+			'Aviser': 'newspaperArticle',
+			'Film': 'videoRecording',
+			'Fjernsyn': 'videoRecording',
+			'Radio': 'radioBroadcast',
+			'Kart': 'map',
+		};
+		var nodes = ZU.xpath(doc, '//input[@id="mediaType"]');
+		if (nodes.length && mediaTypes[nodes[0].value]) {
+			return mediaTypes[nodes[0].value];
+		}
+	}
+}
+
+function doWeb(doc, url) {
+	if (detectWeb(doc, url) == "multiple") {
+		var items = ZU.getItemArray(doc, doc, '/nbsok/nb/');
+		Zotero.selectItems(items, function(items) {
+			if(!items) {
+				return true;
+			}
+			var urls = Object.keys(items);
+			ZU.processDocuments(urls, scrape);
+		});
+	} else {
+		scrape(doc, url);
+	}
+}
+
+function trimBrackets(obj) {
+	// Trim brackets from all values
+	Object.keys(obj).forEach(function(key) {
+		if (typeof(obj[key]) == 'string') {
+			obj[key] = obj[key].replace(/^\[/, '').replace(/\]$/, '');
+		}
+	});
+	return obj;
+}
+
+function extractNumPages(str) {
+	// Borrowed from Library Catalog (PICA). See #756
+	//make sure things like 2 partition don't match, but 2 p at the end of the field do
+	// f., p., and S. are "pages" in various languages
+	// For multi-volume works, we expect formats like:
+	//   x-109 p., 510 p. and X, 106 S.; 123 S.
+	var numPagesRE = /\[?\b((?:[ivxlcdm\d]+[ ,\-]*)+)\]?\s+[fps]\b/ig,
+		numPages = [], m;
+	while(m = numPagesRE.exec(str)) {
+		numPages.push(m[1].trim()
+			.replace(/[ ,\-]+/g,'+')
+			.toLowerCase() // for Roman numerals
+		);
+	}
+	return numPages.join('; ');
+}
+
+function getRIS(url, cb) {
+	ZU.doGet(url, function(text){
+		var translator = Zotero.loadTranslator("import");
+		translator.setTranslator("32d59d2d-b65a-4da4-b0a3-bdd3cfb979e7");
+		translator.setString(text);
+		translator.setHandler("itemDone", function(obj, item) {
+			cb(item);
+		});
+		translator.translate();
+	});
+}
+
+function scrape(doc, url) {
+	var endNote = ZU.xpath(doc, '//a[text()="EndNote"]');
+	getRIS(endNote[0].href, function(item) {
+		item = trimBrackets(item);
+
+		// Normalize notes
+		item.notes.forEach(function(note) {
+			note.note = note.note
+				.replace(/^<p>/, '').replace(/<\/p>$/, '')  // paragraph tags
+				.replace(/&nbsp;/g, ' ')        // hard spaces
+				.replace(/ +/g, ' ')            // multiple spaces
+				;
+		});
+
+		if (item.numPages) {
+			item.numPages = extractNumPages(item.numPages);
+		}
+
+		// Add permalink
+		fields = ZU.getItemArray(doc, doc, 'http://urn.nb.no/URN:NBN');
+		Object.keys(fields).forEach(function(url) {
+			item.url = url;
+		});
```

Comment:
It was just a quick and dirty way to get the link that starts with `http://urn.nb.no/URN:NBN`. This rewrite is probably more efficient and transparent? (I know the link is somewhere inside `#preview_metadata`, so I can narrow in on that.)

```javascript
                var container = doc.getElementById('preview_metadata');
		if (container) {
			var links = container.getElementsByTagName('a');
			for(var i=0; i<links.length; i++) {
				if (links[i].href && links[i].href.indexOf('urn.nb.no/URN:NBN') != -1) {
					item.url = links[i].href;
					break;
				}
			}
		}
````